datafusion/physical-plan/src/statistics.rs

```diff
@@ -277,6 +262,44 @@ impl MinMaxStatistics {
             .zip(self.min_by_sort_order.iter().skip(1))
             .all(|(max, next_min)| max < next_min)
     }
+
+    /// Computes a bin-packing of the min/max rows in these statistics
+    /// into chains, such that elements in a chain are non-overlapping and ordered
+    /// amongst one another.
+    /// This bin-packing is optimal in the sense that it has the fewest number of chains.
+    pub fn first_fit(&self) -> Vec<Vec<usize>> {
```

Comment:
How do we know there are no overlapping ranges here? It seems like we would also have to check if the ranges overlapped and if any did we can't do this packing

This may be checked elsewhere but I didn't see it in a cursory glance

Reply:
If no ranges overlapped, they could all be ordered into a single chain. If some ranges _do_ overlap, they get placed into separate chains. The check for non-overlapping-ness happens in this logic: https://github.com/apache/datafusion/blob/31d371695478839b3112b788ec1966142a6fc0ff/datafusion/physical-plan/src/statistics.rs#L286-L294